src/document.c

```diff
@@ -930,36 +930,89 @@ typedef struct
 } FileData;
 
 
+static time_t get_mtime(const gchar *locale_filename)
+{
+	GError *error = NULL;
+	const gchar *err_msg = NULL;
+	time_t time = 0;
+
+	if (!file_prefs.use_safe_file_saving && file_prefs.use_gio_unsafe_file_saving)
+	{
+		GFile *file = g_file_new_for_path(locale_filename);
+		GFileInfo *info = g_file_query_info(file, G_FILE_ATTRIBUTE_TIME_MODIFIED, G_FILE_QUERY_INFO_NONE, NULL, &error);
+
+		if (info)
+		{
+			GTimeVal timeval;
+
+			g_file_info_get_modification_time(info, &timeval);
+			g_object_unref(info);
+			time = timeval.tv_sec;
+		}
+		else if (error)
+			err_msg = error->message;
+
+		g_object_unref(file);
+	}
+	else
+	{
+		struct stat st;
+
+		if (g_stat(locale_filename, &st) == 0)
+			time = st.st_mtime;
+		else
+			err_msg = g_strerror(errno);
+	}
+
+	if (err_msg)
+	{
+		gchar *utf8_filename = utils_get_utf8_from_locale(locale_filename);
+
+		ui_set_statusbar(TRUE, _("Could not open file %s (%s)"),
+			utf8_filename, err_msg);
+		g_free(utf8_filename);
+	}
+
+	if (error)
+		g_error_free(error);
+
+	return time;
+}
+
+
 /* loads textfile data, verifies and converts to forced_enc or UTF-8. Also handles BOM. */
 static gboolean load_text_file(const gchar *locale_filename, const gchar *display_filename,
 	FileData *filedata, const gchar *forced_enc)
 {
 	GError *err = NULL;
-	struct stat st;
 
 	filedata->data = NULL;
 	filedata->len = 0;
 	filedata->enc = NULL;
 	filedata->bom = FALSE;
 	filedata->readonly = FALSE;
+	filedata->mtime = get_mtime(locale_filename);
 
-	if (g_stat(locale_filename, &st) != 0)
-	{
-		ui_set_statusbar(TRUE, _("Could not open file %s (%s)"),
-			display_filename, g_strerror(errno));
+	if (filedata->mtime == 0)
 		return FALSE;
-	}
 
-	filedata->mtime = st.st_mtime;
+	if (!file_prefs.use_safe_file_saving && file_prefs.use_gio_unsafe_file_saving)
+	{
+		GFile *file = g_file_new_for_path(locale_filename);
+
+		g_file_load_contents(file, NULL, &filedata->data, &filedata->len, NULL, &err);
+		g_object_unref(file);
+	}
+	else
+		g_file_get_contents(locale_filename, &filedata->data, &filedata->len, &err);
 
-	if (! g_file_get_contents(locale_filename, &filedata->data, NULL, &err))
+	if (err)
 	{
 		ui_set_statusbar(TRUE, "%s", err->message);
 		g_error_free(err);
 		return FALSE;
 	}
 
-	filedata->len = (gsize) st.st_size;
```

Comment:
@b4n Maybe I miss something but right now (without my patch):

1. The stat() length value is passed to encodings_convert_to_utf8_auto() here

https://github.com/geany/geany/blob/master/src/document.c#L963

2. encodings_convert_to_utf8_auto() stores this value to buffer.size and strlen() to buffer.len parameter passed to handle_buffer() here:

https://github.com/geany/geany/blob/master/src/encodings.c#L1082

3. Finally, handle_buffer() compares these two and does some magic based on this:

https://github.com/geany/geany/blob/master/src/encodings.c#L1029

The thing is we need both the stat() value and strlen() value in order to do what's done in (3). Right now, my patch discards the stat() value which causes that the buffer->len != buffer->size check is equivalent to FALSE.

In the commit I mentioned the size was just moved into encodings.c but the two values are still there.

So I think this is something that should be fixed in my patch - personally I'd move the stat call inside handle_buffer() because passing the size value from somewhere else is a bit misleading (I at least assumed it was the size of the string which isn't the case).

What do you think?